### ISSUE TITLE
JITSU SOMETIMES CONFUSES COLUMNS

### ISSUE THREAD

**Issue by viggin543** [jitsucom/jitsu#864]
## Environment

- **Deploy method**: k8s  `SPOT INSTANCES`
- **Jitsu version**:  `1.38.4`
- **OS**:  `alpine linux`
- **Cloud Provider**: `GCP`

## Description

sometimes jitsu confuses columns.
for example I see IP addresses in a column that supposed to contain the user name.
This happens not often but causes trouble.
JITSU version:  1.38.4 + snowflake dest

In our case it happened one time out of 1_654_055 events of that group.
( its enough to cause trouble tough... )

## corrupt row example:
```csv
_TIMESTAMP,API_KEY,APPNAME,CONTENTLANGUAGE,DOC_ENCODING,DOC_HOST,DOC_PATH,DOC_SEARCH,E,EVENT_TYPE,EVENTN_CTX_EVENT_ID,FINGERPRINT,ITEM_COUNT,ITEMS,ITEMS_SUBTOTAL_PRICE,LOCAL_TZ_OFFSET,LOCATION_CITY,LOCATION_COUNTRY,LOCATION_LATITUDE,LOCATION_LONGITUDE,LOCATION_REGION,MICROSESSIONCOUNT,MICROSESSIONID,ORIGINAL_TOTAL_PRICE,PAGE_PAGETYPE,PAGE_TITLE,PAGEID,PARSED_UA_OS_FAMILY,PARSED_UA_OS_VERSION,PARSED_UA_UA_FAMILY,PARSED_UA_UA_VERSION,REFERER,S,SCREEN_RESOLUTION,SHOPID,SOURCE_IP,SRC,THEMEID,TOTAL_DISCOUNT,TOTAL_PRICE,TOTAL_WEIGHT,UNIQTOKEN,URL,USE_CASES,USER_AGENT,USER_ANONYMOUS_ID,USER_HASHED_ANONYMOUS_ID,USER_LANGUAGE,UTC_TIME,VISITTOKEN,VP_SIZE,VARIANT_PRICE,SHOPIFY_COUNTRY,IDS_GA,UTM_CAMPAIGN,PRODUCT_TYPE,PRODUCT_ID,CLIENTID,CARTID,LOCATION_ZIP,VARIANT_SKU,PARSED_UA_DEVICE_BRAND,PARSED_UA_DEVICE_FAMILY,PARSED_UA_DEVICE_MODEL,CLICK_ID_GCLID,PAGE_RESOURCEID,QTY,UTM_CONTENT,PARSED_UA_BOT,VARIANTID,CLICK_ID_FBCLID,IDS_FBP,VARIANT_NAME,VARIANT_ID,UTM_SOURCE,PRODUCT_VENDOR,UTM_MEDIUM,UTM_TERM,IDS_YM_UID,LOCATION_COUNTRY_NAME,LOCATION_CONTINENT,FINGERPRINTDATA,TAGNAME,TEXT,LMI_PARAMS_LMI_CLASS,LMI_PARAMS_LMI_TYPE,LMI_PARAMS_LMI_FROM,USE_CASE,ORDER_ID,SUBTOTAL_PRICE,CUSTOMER_ID,TOTAL_TAX,LINE_ITEMS,LMI_CLASS,LMI_POSITION,USE_CASE_VARIANT,FIRST_SEEN,ORDER_CREATED_AT
2022-03-19 11:30:17.378799,js.....sn1bylj0yqt3fipr95vvc7,,,UTF-8,www.karmaandluck.com,/es/collections/mens-karma-red-collection/products/balance-toxic-emotions-red-string-hamsa-evil-eye-charm-bracelet,,,USE_CASE,27487d5d-620a-4fa6-a82d-2d150f34af66,,1,"[{\""id\"":32827215970384,\""price\"":6900,\""sku\"":\""MBJCHM405\"",\""quantity\"":1}]",6900,240,Grand Rapids,US,42.9638,-85.6665,MI,,,6900,product,Brazalete Hilo Rojo Para Balance ,,iOS,15.4,Mobile Safari,15.4, Karma \u0026 Luck,,428x926,5a9002ed-9b41-4669-bdaa-b2ce29889c66,https://www.karmaandluck.com/es/collections/mens-karma-red-collection,jitsu,,0,6900,85,,https://www.karmaandluck.com/es/collections/mens-karma-red-collection/products/balance-toxic-emotions-red-string-hamsa-evil-eye-charm-bracelet,,"Mozilla/5.0 (iPhone; CPU iPhone OS 15_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.4 Mobile/15E148 Safari/604.1",v2jp0yihl9,6a32338de92187331a95c2bb6621db9e,es-US,2022-03-19 11:30:17.140000,,428x746,,,GA1.2.349823307.1647101927,,Bracelets - Red Mens,3622058393680,,5f679b6fe1b1aef15cbd74760dfb5354,49502,,Apple,iPhone,iPhone,,3622058393680,,,,,,fb.1.1647101927050.1796683760,,,,Laursain,,,,United States,North America,,,,,,,50a842f5-a645-4b59-9249-b36b469d78a6,,,,,,,,104.28.76.161,1647101925,

```
The column that contain an incorrect value here is `USE_CASE_VARIANT`  (third from the end) it contains an IP address instead of an id
it also seems like a column name `USE_CASE` appears as a value here.

## valid row example
```csv
_TIMESTAMP,API_KEY,APPNAME,CONTENTLANGUAGE,DOC_ENCODING,DOC_HOST,DOC_PATH,DOC_SEARCH,E,EVENT_TYPE,EVENTN_CTX_EVENT_ID,FINGERPRINT,ITEM_COUNT,ITEMS,ITEMS_SUBTOTAL_PRICE,LOCAL_TZ_OFFSET,LOCATION_CITY,LOCATION_COUNTRY,LOCATION_LATITUDE,LOCATION_LONGITUDE,LOCATION_REGION,MICROSESSIONCOUNT,MICROSESSIONID,ORIGINAL_TOTAL_PRICE,PAGE_PAGETYPE,PAGE_TITLE,PAGEID,PARSED_UA_OS_FAMILY,PARSED_UA_OS_VERSION,PARSED_UA_UA_FAMILY,PARSED_UA_UA_VERSION,REFERER,S,SCREEN_RESOLUTION,SHOPID,SOURCE_IP,SRC,THEMEID,TOTAL_DISCOUNT,TOTAL_PRICE,TOTAL_WEIGHT,UNIQTOKEN,URL,USE_CASES,USER_AGENT,USER_ANONYMOUS_ID,USER_HASHED_ANONYMOUS_ID,USER_LANGUAGE,UTC_TIME,VISITTOKEN,VP_SIZE,VARIANT_PRICE,SHOPIFY_COUNTRY,IDS_GA,UTM_CAMPAIGN,PRODUCT_TYPE,PRODUCT_ID,CLIENTID,CARTID,LOCATION_ZIP,VARIANT_SKU,PARSED_UA_DEVICE_BRAND,PARSED_UA_DEVICE_FAMILY,PARSED_UA_DEVICE_MODEL,CLICK_ID_GCLID,PAGE_RESOURCEID,QTY,UTM_CONTENT,PARSED_UA_BOT,VARIANTID,CLICK_ID_FBCLID,IDS_FBP,VARIANT_NAME,VARIANT_ID,UTM_SOURCE,PRODUCT_VENDOR,UTM_MEDIUM,UTM_TERM,IDS_YM_UID,LOCATION_COUNTRY_NAME,LOCATION_CONTINENT,FINGERPRINTDATA,TAGNAME,TEXT,LMI_PARAMS_LMI_CLASS,LMI_PARAMS_LMI_TYPE,LMI_PARAMS_LMI_FROM,USE_CASE,ORDER_ID,SUBTOTAL_PRICE,CUSTOMER_ID,TOTAL_TAX,LINE_ITEMS,LMI_CLASS,LMI_POSITION,USE_CASE_VARIANT,FIRST_SEEN,ORDER_CREATED_AT
2022-01-28 00:02:49.789784,js.....sn1bylj0yqt3fipr95vvc7,storefront,en,UTF-8,www.karmaandluck.com,/collections/mens-jewelry/products/supreme-manifestation-black-onyx-heishi-mantra-bracelet,,,USE_CASE,3e0e5083-561c-4739-8d04-df8f85ee070d,,0,[],0,420,Denver,US,39.7301,-104.9078,CO,3,9dfdeaaa-5739-4385-DC8F-84BE7ABDE0F8,0,product,Supreme Manifestation - Black Onyx Heishi Mantra Bracelet | Karma and Luck,9dfdeb21-B400-4EFD-C39B-1FAC272C50CE,iOS,15.1,Facebook,,https://www.karmaandluck.com/collections/mens-jewelry,,390x844,23182543,174.198.128.100,jitsu,120652267600,0,0,0,3f7edda5-fc05-4bd4-9832-249f9f2030b4,https://www.karmaandluck.com/collections/mens-jewelry/products/supreme-manifestation-black-onyx-heishi-mantra-bracelet,"[{\""name\"":\""50a842f5-a645-4b59-9249-b36b469d78a6\"",\""variant\"":\""b432fcf4953663b7\""}]","Mozilla/5.0 (iPhone; CPU iPhone OS 15_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/19B74 [FBAN/FBIOS;FBDV/iPhone13,3;FBMD/iPhone;FBSN/iOS;FBSV/15.1;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]",3tyd5fwmir,e475bf5f77d9310b69b435f51e51f1ae,en-US,2022-01-28 00:02:49.734000,5ce9afcc-7df4-42f8-a98d-355f537af37d,390x655,,,GA1.2.237402473.1642726123,,Bracelet,6565906153552,,,80220,,Apple,iPhone,"iPhone13,3",,6565906153552,,,,,,fb.1.1642726123410.1104310612,,,,Yasmin,,,,United States,North America,,,,,,,50a842f5-a645-4b59-9249-b36b469d78a6,,,,,,,,b432fcf4953663b7,1643298000,

```

## Steps to reproduce

send data to jitsu

## Expected behavior

should not confuse columns

## Actual behavior

Very rearly jitsu confuses columns and creates corrupted rows that can screw analytics data

**Comment by absorbb** [comment]
Hi! @viggin543 
Apparently every SQL query that jitsu runs on database has randomized order of columns.
I have a theory that such behavior may not play well with query optimizers on some DBs.
I need more time to investigate. Will get back to this ticket after we make 1.41.0 release tomorrow

**Comment by viggin543** [comment]
🚀

**Comment by absorbb** [comment]
@viggin543 
Can you please send Jitsu log message related to this error?

**Comment by absorbb** [comment]
@viggin543 
also do you use stream mode or batch?

**Comment by viggin543** [comment]
@absorbb  BATCH

**Comment by viggin543** [comment]
it's not an error.. everything worked.
Just the data was written in a corrupted way.

An IP address was written to the wrong column.
We don't send custom data that contain an IP address explicitly. So something in jitsu backend caused it.

we also run on spot instances, that can be terminated in mid-process. 
@absorbb

**Comment by absorbb** [comment]
Thank for the info.
SOURCE_IP is added on Jitsu backend.
It looks like that many columns are in the wrong places in the first example.

I've build the new jitsu image with tag `sf_fix` based on 1.41.0
which additionally have fixes for column order in SQL queries
and goshowflake driver updated to the latest version 1.6.8 

is it possible for you to try it on your setup?

**Comment by viggin543** [comment]
@absorbb 

Is it safe for production?
we didn't experience this phenomenon on staging env ( probably because it has almost no traffic )

On prod, we have quite some traffic, and we observed this behavior a couple of times ( each time deleting corrupted rows manually ) resolved our problems.

**Comment by absorbb** [comment]
@viggin543 
>Is it safe for production?
we are going to test it internally. Next week it will be safe to say

**Comment by absorbb** [comment]
@viggin543 
few more questions:
what stage do you use: s3 or google cloud storage?
do you use user recognition?

**Comment by viggin543** [comment]
@absorbb 
GCS, user recognition disabled

**Comment by absorbb** [comment]
@viggin543 
Thank you!

Found the reason. 
Spanish version of your site uses || delimiter for page titles. e.g.: https://www.karmaandluck.com/es/collections/mens-karma-red-collection/products/balance-toxic-emotions-red-string-hamsa-evil-eye-charm-bracelet
The same delimiter Jitsu use as CSV separator when we store files to GCS. That breaks column structure for events from such pages.
That is bug that we going to fix asap. 


Also during testing i somehow got redirected to https://de-karmaandluck.glopalstore.com/collections/mens-karma-red-collection/products/balance-toxic-emotions-red-string-hamsa-evil-eye-charm-bracelet
where your jitsu setup doesnt work due to CORS issue. (probably because of different domain) That may be something that you may want to fix.

**Comment by viggin543** [comment]
@absorbb 
No thank you!

All those delimiters in CSV-like formats are a crawling nest of bags.

Top reasons for software bugs:
- timestamps
- CSV delimiters 

:)

Thanks for the CORS issue catch.

**Comment by absorbb** [comment]
Fixed in [1.41.1](https://github.com/jitsucom/jitsu/releases/tag/v1.41.1)
